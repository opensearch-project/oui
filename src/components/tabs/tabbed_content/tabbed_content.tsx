/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { Component, createRef, HTMLAttributes, ReactNode } from 'react';

import { htmlIdGenerator } from '../../../services';

import { OuiTabs, OuiTabsDisplaySizes, OuiTabsSizes } from '../tabs';
import { OuiTab } from '../tab';
import { CommonProps } from '../../common';

/**
 * Marked as const so type is `['initial', 'selected']` instead of `string[]`
 */
export const AUTOFOCUS = ['initial', 'selected'] as const;

export interface OuiTabbedContentTab {
  id: string;
  name: ReactNode;
  content: ReactNode;
}

interface OuiTabbedContentState {
  selectedTabId: string | undefined;
  preservedTabsId: string[];
  inFocus: boolean;
}

export type OuiTabbedContentProps = CommonProps &
  Omit<HTMLAttributes<HTMLDivElement>, 'autoFocus'> & {
    /**
     * When tabbing into the tabs, set the focus on `initial` for the first tab,
     * or `selected` for the currently selected tab. Best use case is for inside of
     * overlay content like popovers or flyouts.
     */
    autoFocus?: 'initial' | 'selected';
    /**
     * Choose `default` or alternative `condensed` display styles
     */
    display?: OuiTabsDisplaySizes;
    /**
     * Evenly stretches each tab to fill the horizontal space
     */
    expand?: boolean;
    /**
     * Use this prop to set the initially selected tab while letting the tabbed content component
     * control selection state internally
     */
    initialSelectedTab?: OuiTabbedContentTab;
    onTabClick?: (selectedTab: OuiTabbedContentTab) => void;
    /**
     * Use this prop if you want to control selection state within the owner component
     */
    selectedTab?: OuiTabbedContentTab;
    size?: OuiTabsSizes;
    /**
     * Each tab needs id and content properties, so we can associate it with its panel for accessibility.
     * The name property (a node) is also required to display to the user.
     */
    tabs: OuiTabbedContentTab[];
    /**
     * Use this prop to allow tab content stay in dom, if the tab has been selected.
     */
    preserveTabContent?: boolean;
  };

export class OuiTabbedContent extends Component<
  OuiTabbedContentProps,
  OuiTabbedContentState
> {
  static defaultProps: Partial<OuiTabbedContentProps> = {
    autoFocus: 'initial',
    preserveTabContent: false,
  };

  private readonly rootId = htmlIdGenerator()();

  private readonly tabsRef = createRef<HTMLDivElement>();

  constructor(props: OuiTabbedContentProps) {
    super(props);

    const { initialSelectedTab, selectedTab, tabs } = props;

    // Only track selection state if it's not controlled externally.
    let selectedTabId;
    if (!selectedTab) {
      selectedTabId =
        (initialSelectedTab && initialSelectedTab.id) || tabs[0].id;
    }

    const preservedTabsId = [];
    preservedTabsId.push(
      selectedTab
        ? selectedTab.id
        : (initialSelectedTab && initialSelectedTab.id) || tabs[0].id
    );

    this.state = {
      selectedTabId,
      preservedTabsId,
      inFocus: false,
    };
  }

  componentDidMount() {
    // IE11 doesn't support the `relatedTarget` event property for blur events
    // but does add it for focusout. React doesn't support `onFocusOut` so here we are.
    if (this.tabsRef.current) {
      // Current short-term solution for event listener (see https://github.com/elastic/eui/pull/2717)
      this.tabsRef.current.addEventListener(
        'focusout' as 'blur',
        this.removeFocus
      );
    }
  }

  componentWillUnmount() {
    if (this.tabsRef.current) {
      // Current short-term solution for event listener (see https://github.com/elastic/eui/pull/2717)
      this.tabsRef.current.removeEventListener(
        'focusout' as 'blur',
        this.removeFocus
      );
    }
  }

  focusTab = () => {
    const targetTab: HTMLDivElement | null = this.tabsRef.current!.querySelector(
      `#${this.state.selectedTabId}`
    );
    targetTab!.focus();
  };

  initializeFocus = () => {
    if (!this.state.inFocus && this.props.autoFocus === 'selected') {
      // Must wait for setState to finish before calling `.focus()`
      // as the focus call triggers a blur on the first tab
      this.setState({ inFocus: true }, () => {
        this.focusTab();
      });
    }
  };

  // todo: figure out type for blurEvent
  removeFocus = (blurEvent: FocusEvent) => {
    // only set inFocus to false if the wrapping div doesn't contain the now-focusing element
    const currentTarget = blurEvent.currentTarget! as HTMLElement;
    const relatedTarget = blurEvent.relatedTarget! as HTMLElement;
    if (currentTarget.contains(relatedTarget) === false) {
      this.setState({
        inFocus: false,
      });
    }
  };

  onTabClick = (selectedTab: OuiTabbedContentTab) => {
    const { onTabClick, selectedTab: externalSelectedTab } = this.props;

    if (onTabClick) {
      onTabClick(selectedTab);
    }

    // Only track selection state if it's not controlled externally.
    this.setState((prevState) => ({
      selectedTabId: externalSelectedTab
        ? prevState.selectedTabId
        : selectedTab.id,
      preservedTabsId: [...prevState.preservedTabsId, selectedTab.id],
    }));
  };

  render() {
    const {
      className,
      display,
      expand,
      initialSelectedTab,
      onTabClick,
      selectedTab: externalSelectedTab,
      size,
      tabs,
      autoFocus,
      preserveTabContent,
      ...rest
    } = this.props;

    // Allow the consumer to control tab selection.
    const selectedTab =
      externalSelectedTab ||
      tabs.find(
        (tab: OuiTabbedContentTab) => tab.id === this.state.selectedTabId
      );

    const { content: selectedTabContent, id: selectedTabId } = selectedTab!;

    return (
      <div className={className} {...rest}>
        <OuiTabs
          ref={this.tabsRef}
          expand={expand}
          display={display}
          size={size}
          onFocus={this.initializeFocus}>
          {tabs.map((tab: OuiTabbedContentTab) => {
            const {
              id,
              name,
              content, // eslint-disable-line no-unused-vars
              ...tabProps
            } = tab;
            const props = {
              key: id,
              id,
              ...tabProps,
              onClick: () => this.onTabClick(tab),
              isSelected: tab === selectedTab,
              'aria-controls': `${this.rootId}`,
            };

            return <OuiTab {...props}>{name}</OuiTab>;
          })}
        </OuiTabs>

        {!preserveTabContent && (
          <div
            role="tabpanel"
            id={`${this.rootId}`}
            aria-labelledby={selectedTabId}>
            {selectedTabContent}
          </div>
        )}

        {preserveTabContent &&
          tabs.map((tab: OuiTabbedContentTab) => {
            const { id, content } = tab;

            if (this.state.preservedTabsId.includes(id)) {
              return (
                <div
                  key={id}
                  role="tabpanel"
                  id={id}
                  aria-labelledby={id}
                  style={
                    id === selectedTabId ? undefined : { display: 'none' }
                  }>
                  {content}
                </div>
              );
            }
          })}
      </div>
    );
  }
}
