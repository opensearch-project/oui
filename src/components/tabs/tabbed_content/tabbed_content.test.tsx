/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { requiredProps } from '../../../test';

import { OuiTabbedContent, AUTOFOCUS } from './tabbed_content';

const elasticsearchTab = {
  id: 'es',
  name: 'Elasticsearch',
  content: <p>Elasticsearch content</p>,
};

const kibanaTab = {
  id: 'kibana',
  name: <strong>Kibana</strong>,
  'data-test-subj': 'kibanaTab',
  content: <p>Kibana content</p>,
};

const tabs = [elasticsearchTab, kibanaTab];

describe('OuiTabbedContent', () => {
  test('is rendered with required props and tabs', () => {
    const { container } = render(
      <OuiTabbedContent {...requiredProps} tabs={tabs} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('props', () => {
    describe('onTabClick', () => {
      test('is called when a tab is clicked', () => {
        const onTabClickHandler = jest.fn();
        const { getByTestId } = render(
          <OuiTabbedContent onTabClick={onTabClickHandler} tabs={tabs} />
        );

        fireEvent.click(getByTestId('kibanaTab'));
        expect(onTabClickHandler).toHaveBeenCalledTimes(1);
        expect(onTabClickHandler).toHaveBeenCalledWith(kibanaTab);
      });
    });

    describe('selectedTab', () => {
      test('renders a selected tab', () => {
        const { container } = render(
          <OuiTabbedContent selectedTab={kibanaTab} tabs={tabs} />
        );
        expect(container.firstChild).toMatchSnapshot();
      });
    });

    describe('initialSelectedTab', () => {
      test('renders a selected tab', () => {
        const { container } = render(
          <OuiTabbedContent initialSelectedTab={kibanaTab} tabs={tabs} />
        );
        expect(container.firstChild).toMatchSnapshot();
      });
    });

    describe('size', () => {
      test('can be small', () => {
        const { container } = render(<OuiTabbedContent size="s" tabs={tabs} />);
        expect(container.firstChild).toMatchSnapshot();
      });
    });

    describe('display', () => {
      test('can be condensed', () => {
        const { container } = render(
          <OuiTabbedContent display="condensed" tabs={tabs} />
        );
        expect(container.firstChild).toMatchSnapshot();
      });
    });

    describe('autoFocus', () => {
      AUTOFOCUS.forEach((focusType) => {
        test(`${focusType} is rendered`, () => {
          const { container } = render(
            <OuiTabbedContent autoFocus={focusType} tabs={tabs} />
          );

          expect(container.firstChild).toMatchSnapshot();
        });
      });
    });

    describe('cacheContent', () => {
      test('content of tabs that has been selected before should stay in dom', () => {
        const { container } = render(
          <OuiTabbedContent preserveTabContent={true} tabs={tabs} />
        );
        expect(container.querySelectorAll('div[role="tabpanel"]').length).toBe(
          1
        );

        const kibanaTabButton = container.querySelector(
          'button[id="kibana"]'
        ) as HTMLButtonElement;
        fireEvent.click(kibanaTabButton);
        expect(container.querySelectorAll('div[role="tabpanel"]').length).toBe(
          2
        );
      });
    });
  });

  describe('behavior', () => {
    test("when selected tab state isn't controlled by the owner, select the first tab by default", () => {
      const { container } = render(<OuiTabbedContent tabs={tabs} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('when uncontrolled, the selected tab should update if it receives new content', () => {
      const tabs = [
        elasticsearchTab,
        {
          ...kibanaTab,
        },
      ];
      const { container, rerender } = render(<OuiTabbedContent tabs={tabs} />);

      const kibanaTabButton = container.querySelector(
        'button[id="kibana"]'
      ) as HTMLButtonElement;
      fireEvent.click(kibanaTabButton);

      rerender(
        <OuiTabbedContent
          tabs={[
            elasticsearchTab,
            {
              ...kibanaTab,
              content: <p>updated Kibana content</p>,
            },
          ]}
        />
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
