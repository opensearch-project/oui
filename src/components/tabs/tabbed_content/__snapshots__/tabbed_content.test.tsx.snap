// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiTabbedContent behavior when selected tab state isn't controlled by the owner, select the first tab by default 1`] = `
<div>
  <div
    class="ouiTabs"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="es"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Elasticsearch content
    </p>
  </div>
</div>
`;

exports[`OuiTabbedContent behavior when uncontrolled, the selected tab should update if it receives new content 1`] = `
<OuiTabbedContent
  autoFocus="initial"
  tabs={
    [
      {
        "content": <p>
          Elasticsearch content
        </p>,
        "id": "es",
        "name": "Elasticsearch",
      },
      {
        "content": <p>
          updated Kibana content
        </p>,
        "data-test-subj": "kibanaTab",
        "id": "kibana",
        "name": <strong>
          Kibana
        </strong>,
      },
    ]
  }
>
  <div>
    <OuiTabs
      onFocus={[Function]}
    >
      <div
        className="ouiTabs"
        onFocus={[Function]}
        role="tablist"
      >
        <OuiTab
          aria-controls="generated-id"
          id="es"
          isSelected={false}
          key="es"
          onClick={[Function]}
        >
          <button
            aria-controls="generated-id"
            aria-selected={false}
            className="ouiTab"
            disabled={false}
            id="es"
            onClick={[Function]}
            role="tab"
            type="button"
          >
            <span
              className="ouiTab__content"
            >
              Elasticsearch
            </span>
          </button>
        </OuiTab>
        <OuiTab
          aria-controls="generated-id"
          data-test-subj="kibanaTab"
          id="kibana"
          isSelected={true}
          key="kibana"
          onClick={[Function]}
        >
          <button
            aria-controls="generated-id"
            aria-selected={true}
            className="ouiTab ouiTab-isSelected"
            data-test-subj="kibanaTab"
            disabled={false}
            id="kibana"
            onClick={[Function]}
            role="tab"
            type="button"
          >
            <span
              className="ouiTab__content"
            >
              <strong>
                Kibana
              </strong>
            </span>
          </button>
        </OuiTab>
      </div>
    </OuiTabs>
    <div
      aria-labelledby="kibana"
      id="generated-id"
      role="tabpanel"
    >
      <p>
        updated Kibana content
      </p>
    </div>
  </div>
</OuiTabbedContent>
`;

exports[`OuiTabbedContent is rendered with required props and tabs 1`] = `
<div
  aria-label="aria-label"
  class="testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    class="ouiTabs"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="es"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Elasticsearch content
    </p>
  </div>
</div>
`;

exports[`OuiTabbedContent props autoFocus initial is rendered 1`] = `
<div>
  <div
    class="ouiTabs"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="es"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Elasticsearch content
    </p>
  </div>
</div>
`;

exports[`OuiTabbedContent props autoFocus selected is rendered 1`] = `
<div>
  <div
    class="ouiTabs"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="es"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Elasticsearch content
    </p>
  </div>
</div>
`;

exports[`OuiTabbedContent props display can be condensed 1`] = `
<div>
  <div
    class="ouiTabs ouiTabs--condensed"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="es"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Elasticsearch content
    </p>
  </div>
</div>
`;

exports[`OuiTabbedContent props initialSelectedTab renders a selected tab 1`] = `
<div>
  <div
    class="ouiTabs"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="kibana"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Kibana content
    </p>
  </div>
</div>
`;

exports[`OuiTabbedContent props selectedTab renders a selected tab 1`] = `
<div>
  <div
    class="ouiTabs"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="kibana"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Kibana content
    </p>
  </div>
</div>
`;

exports[`OuiTabbedContent props size can be small 1`] = `
<div>
  <div
    class="ouiTabs ouiTabs--small"
    role="tablist"
  >
    <button
      aria-controls="generated-id"
      aria-selected="true"
      class="ouiTab ouiTab-isSelected"
      id="es"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        Elasticsearch
      </span>
    </button>
    <button
      aria-controls="generated-id"
      aria-selected="false"
      class="ouiTab"
      data-test-subj="kibanaTab"
      id="kibana"
      role="tab"
      type="button"
    >
      <span
        class="ouiTab__content"
      >
        <strong>
          Kibana
        </strong>
      </span>
    </button>
  </div>
  <div
    aria-labelledby="es"
    id="generated-id"
    role="tabpanel"
  >
    <p>
      Elasticsearch content
    </p>
  </div>
</div>
`;
